woocommerce/woocommerce
# Add Default Fallback Marketing Channel Recommendations

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Closes issue https://github.com/woocommerce/woocommerce/issues/43032.

In this PR, we add default fallback data for marketing channel recommendations. In relation to that, we also make changes to the code logic (code refactoring) to make use of the Data Source Poller in the code base, so that it is consistent with other features such as payment gateway suggestions, remote free extensions in onboarding wizard flow etc.

**Code references that we use in this PR:**

- https://github.com/woocommerce/woocommerce/tree/a9550d8894ea5de1f02fc954ff41db8eac487f73/plugins/woocommerce/src/Internal/Admin/RemoteFreeExtensions
- https://github.com/woocommerce/woocommerce/tree/a9550d8894ea5de1f02fc954ff41db8eac487f73/plugins/woocommerce/src/Admin/Features/PaymentGatewaySuggestions
- https://github.com/woocommerce/woocommerce/tree/a9550d8894ea5de1f02fc954ff41db8eac487f73/plugins/woocommerce/src/Internal/Admin/WCPayPromotion
- https://github.com/woocommerce/woocommerce/tree/a9550d8894ea5de1f02fc954ff41db8eac487f73/plugins/woocommerce/src/Admin/Features/ShippingPartnerSuggestions

**Technical notes on the changes:**

Based on the above code references, I created a new "MarketingRecommendations" feature that replaces the existing "MarketingSpecs". The new feature makes use of the DataSourcePoller class which retrieves recommendation data from woocommerce.com API when the marketing suggestions is turned on, and retrieves recommendation data generated from within WooCommerce plugin (without making request to woocommerce.com API) when the marketing suggestions is turned off. 

With the data source poller, after it retrieves recommendation data from woocommerce.com API, the data will be stored in a transient called `woocommerce_admin_marketing_recommendations_specs`. The transient name is generated by the DataSourcePoller in https://github.com/woocommerce/woocommerce/blob/ccdfb16fc881f56e68d59e9d34810e53841126c7/plugins/woocommerce/src/Admin/DataSourcePoller.php#L68

The old transient `wc_marketing_recommended_plugins` will not be used anymore.

In `DefaultMarketingRecommendations`, most of the code comes from the woocommerce.com API code base. The idea is that when we make changes to the recommendations API in woocommerce.com, we should be able to quickly copy and paste the code change into our WooCommerce plugin code base here with minimal changes.

I also added new icons for the recommended extensions. Some of these icons are not currently being used (e.g. those icons that start with "woocommerce-"), but we want to make it the same as woocommerce.com API and they are being referenced in `DefaultMarketingRecommendations`. In addition, with this in place, in the future we can make default recommendations  show up in Marketing > Coupons page, even when marketing suggestion feature is turned off. 

**Note for next PR:**

Based on the current timeline, this feature shall be released in WooCommerce 8.6.

I will create a separate PR with a separate test instructions to remove the old transient `wc_marketing_recommended_plugins` when the plugin is upgraded from WooCommerce 8.5 to 8.6. 

**Screenshot:**

![image](https://github.com/woocommerce/woocommerce/assets/417342/054858eb-bab7-4ff1-bb27-7b3dd37c023f)



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

With marketplace suggestions turned on:

1. Enable marketplace suggestions in WooCommerce Settings > Advanced > Woo.com: `/wp-admin/admin.php?page=wc-settings&tab=advanced&section=woocommerce_com`
2. Check that there isn't a transient called `woocommerce_admin_marketing_recommendations_specs`. If there is, delete it. 
3. If there is a transient called `wc_marketing_recommended_plugins`, delete it.
4. Go to Marketing page: `/wp-admin/admin.php?page=wc-admin&path=%2Fmarketing`
5. We should see marketing channel recommendations. The recommendation data corresponds to the data retrieved from `https://woocommerce.com/wp-json/wccom/marketing-tab/1.3/recommendations.json`.
6. Check that the data is stored in a transient called `woocommerce_admin_marketing_recommendations_specs`. 
7. Check that there is no transient called `wc_marketing_recommended_plugins`.

With marketplace suggestions turned off:

1. Disable marketplace suggestions in WooCommerce Settings > Advanced > Woo.com: `/wp-admin/admin.php?page=wc-settings&tab=advanced&section=woocommerce_com`
2. Check that there isn't a transient called `woocommerce_admin_marketing_recommendations_specs`. If there is, delete it.
3. If there is a transient called `wc_marketing_recommended_plugins`, delete it.
4. Go to Marketing page: `/wp-admin/admin.php?page=wc-admin&path=%2Fmarketing`
5. We should see marketing channel recommendations. The recommendation data corresponds to the data generated in `DefaultMarketingRecommendations.php` file in this PR, with `categories` array containing `'marketing'` and `subcategories` array containing the `'sales-channels'` array. 
    - Note that other subcategories will not be shown in the page because the "Discover more marketing tools" card is not shown when the marketing suggestions feature is turned off. 
7. Check that there isn't a transient called `woocommerce_admin_marketing_recommendations_specs`. 
8. Check that there is no transient called `wc_marketing_recommended_plugins`.

Code base check:

1. Check that the removed code in `MarketingSpecs.php` file are not being used in other places - if they are, they might cause run time errors.



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

Add default fallback marketing channel and tool recommendations.

#### Comment 

</details>
